Extend geometry robustness testing

sethrj opened the pull request:
This adds new testing utilities for ensuring the safety and boundary intercept distances are consistent. It updates the VecGeom version to v1.2.5 to fix the last two remaining known tracking issues which are documented below (and in the follow-up comments).

----

Follow-on to #844: with @amandalund 's help I was able to find some troublesome points during tracking through cms2018. I've added an excerpt (physical volume `SF0x7f4a8f121340`) from that geometry as a unit test, found a place where reinitializing in the middle of the volume *changed* the logical volume, and confirmed that Geant4's navigator does *not* have the same behavior.

@agheata can you try reproducing this failure, using https://github.com/sethrj/celeritas/blob/vecgeom-sf-failure/test/celeritas/data/sf.gdml , by starting at `-23.235161088317067, -48.92837314875102, 2.485790202725453` along `-0.6431446435132148, -0.7270389178900708, 0.2403734165699165`, crossing into the next volume, moving halfway to the next volume distance, and checking that `BVHNavigator::LocatePointIn` gives a consistent state before and after?

```
reinitialization changed the volume at {-23.578149423779067,-49.316102209046122,2.6139810830665646} along {-0.64314464351321488,-0.72703891789007091,0.24037341656991656} from SFBY to SF (alleged safety: 0.10499999999999991)
Google Test trace:
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:874: Trouble
/Users/seth/.local/src/celeritas/test/celeritas/GenericGeoTestBase.cc:351: Failure
Value of: next.distance / 2
  Actual: 0.53239576760977803
Expected: new_next.distance
Which is: 3.4116400205757862
(Relative error -0.84394726161058309 exceeds tolerance 1e-10)
Google Test trace:
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:874: Trouble
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:882: Failure
Values in: result.volumes
 Expected: expected_volumes
1 of 3 elements differ
i expected_volumes   result.volumes
1          "SFBY"       "SFBY/SF"

Google Test trace:
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:874: Trouble
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:885: Failure
Values in: result.distances
 Expected: expected_distances
1 of 3 elements differ
by 9.9999999999999998e-13 relative error or 1e-14 absolute error
i  expected_distances   result.distances         Difference
2     1.2305697838786   0.452742755915399  -0.632086890279062

Google Test trace:
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:874: Trouble
/Users/seth/.local/src/celeritas/test/celeritas/ext/Vecgeom.test.cc:888: Failure
Values in: result.halfway_safeties
 Expected: expected_hw_safety
2 of 3 elements differ
by 9.9999999999999998e-13 relative error or 1e-14 absolute error
i  expected_hw_safety             ACTUAL         Difference
0 8.9059349464482e-05 8.90593494646402e-05 1.7766327229934e-12
2    0.12134753426698   0.145576978128889   0.199669849150797
```

sethrj committed:
Add a field test base class

sethrj committed:
Add checked geo track view

sethrj committed:
Fix failing ORANGE geometry

sethrj committed:
Warn instead of crash on zero safety change

sethrj committed:
Fix edge case in Geant field propagation

sethrj committed:
Add CMSE polycones and propagation heuristic test

sethrj committed:
Add test with incorrect VG safeties for CMSE

sethrj committed:
Add CMSE heuristic test

sethrj committed:
Add failing raytrace through sf

sethrj committed:
Minimize failure

agheata commented:
Hi @sethrj and @amandalund , sorry for taking a bit longer, here it is, done both with `BVHNavigator` and `LoopNavigator`. Navigation looks consistent, although I did a few mistakes in the checking code before getting it right, so I'm attaching both the code and the output. 

One thing I did wrong before confirming the location of the mid-point was to forget to clear the navigation state object. Another thing is that I forgot that our `ComputeStateAndNextVolume` computes the next hit volume but WITHOUT relocating, so `RelocateToNextVolume` is a mandatory call to find the actual next state. I also print points/distances in cm now and with increased precision, to be consistent with your units.

The code:
```cpp
template <typename Navigator>
void TestPoint(vecgeom::Vector3D<double> const &point, vecgeom::Vector3D<double> const &dir)
{
  // Direction MUST be normalized
  assert(dir.IsNormalized());
  vecgeom::NavStateIndex navState, nextState;
  auto world = vecgeom::GeoManager::Instance().GetWorld();
  // The locator assumes below navState being cleared
  Navigator::LocatePointIn(world, point, navState, true);
  auto top          = navState.Top();
  int id            = top ? top->id() : 0;
  int logic_id      = top ? top->GetLogicalVolume()->id() : 0;
  std::string vname = top ? top->GetName() : "outside";
  std::cout << typeid(Navigator).name()
            << "_______________________________________________________________________________________\n";
  std::cout << std::fixed << std::setprecision(15) << "point: " << point / copcore::units::cm << " [cm] dir: " << dir
            << "  located in volume: " << vname << " pvol_id = " << id << " logic_id = " << logic_id << " \nstate: ";
  navState.Print();
  auto safety             = Navigator::ComputeSafety(point, navState);
  auto geometryStepLength = Navigator::ComputeStepAndNextVolume(point, dir, 1.e30, navState, nextState, 0.);
  assert(safety <= geometryStepLength && "safety is not safe");
  std::cout << "safety  = " << safety / copcore::units::cm
            << " [cm] distance = " << geometryStepLength / copcore::units::cm << " [cm]\n";
  // Propagate on boundary
  auto point_on_boundary = vecgeom::Vector3D<double>(point + geometryStepLength * dir);
  std::cout << "propagated point on boundary: " << point_on_boundary / copcore::units::cm << " [cm]\n";
  if (nextState.Top() != nullptr) {
    // Relocation assumes nextState to be seeded by ComputeStepAndNextVolume
    Navigator::RelocateToNextVolume(point_on_boundary, dir, nextState);
    top      = nextState.Top();
    id       = top ? top->id() : 0;
    logic_id = top ? top->GetLogicalVolume()->id() : 0;
    vname    = top ? top->GetName() : "outside";
    std::cout << "Relocated volume: " << vname << " pvol_id = " << id << " logic_id = " << logic_id << " \nstate: ";
    nextState.Print();
    // Make next state current, to be able to compute the step in the next volume
    navState           = nextState;
    geometryStepLength = Navigator::ComputeStepAndNextVolume(point_on_boundary, dir, 1.e30, navState, nextState, 0.);
    std::cout << "Distance from boundary point: " << geometryStepLength / copcore::units::cm << " [cm]\n";
    if (!nextState.IsOutside()) {
      auto point_middle = vecgeom::Vector3D<double>(point_on_boundary + 0.5 * geometryStepLength * dir);
      std::cout << "point in the middle before next crossing: " << point_middle / copcore::units::cm << " [cm]\n";
      // Uste the global locatior to confirm the geometry state, now contained in navState.
      // State MUST be cleared before locating starting from top
      nextState.Clear();
      BVHNavigator::LocatePointIn(world, point_middle, nextState, true);
      top      = nextState.Top();
      id       = top ? top->id() : 0;
      logic_id = top ? top->GetLogicalVolume()->id() : 0;
      vname    = top ? top->GetName() : "outside";
      std::cout << "Volume at mid-distance: " << vname << " pvol_id = " << id << " logic_id = " << logic_id
                << " \nstate: ";
      nextState.Print();
      assert(navState.GetNavIndex() == nextState.GetNavIndex());
      std::cout << " consistency confirmed\n";
    }
  }
}
```

The output:
```console
N3cxx12BVHNavigatorE_______________________________________________________________________________________
point: (-23.235161088317067, -48.928373148751021, 2.485790202725453) [cm] dir: (-0.643144643513215, -0.727038917890071, 0.240373416569917)  located in volume: SFLY1a0x7f4a8effa000 pvol_id = 886 logic_id = 313 
state: navInd=109148, id=3520, level=1/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000>
safety  = 0.000178118698929 [cm] distance = 0.000903139442303 [cm]
propagated point on boundary: (-23.235741937611728, -48.929029766273850, 2.486007293438839) [cm]
Relocated volume: SFBY0x7f4a8f3836c0 pvol_id = 391 logic_id = 308 
state: navInd=109644, id=3536, level=2/3,  onBoundary=true, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000/SFBY0x7f4a8f3836c0>
Distance from boundary point: 1.064791535219556 [cm]
point in the middle before next crossing: (-23.578149423779063, -49.316102209046122, 2.613981083066565) [cm]
Volume at mid-distance: SFBY0x7f4a8f3836c0 pvol_id = 391 logic_id = 308 
state: navInd=109644, id=3536, level=2/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000/SFBY0x7f4a8f3836c0>
 consistency confirmed
N3cxx13LoopNavigatorE_______________________________________________________________________________________
point: (-23.235161088317067, -48.928373148751021, 2.485790202725453) [cm] dir: (-0.643144643513215, -0.727038917890071, 0.240373416569917)  located in volume: SFLY1a0x7f4a8effa000 pvol_id = 886 logic_id = 313 
state: navInd=109148, id=3520, level=1/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000>
safety  = 0.000178118698929 [cm] distance = 0.000903139442303 [cm]
propagated point on boundary: (-23.235741937611728, -48.929029766273850, 2.486007293438839) [cm]
Relocated volume: SFBY0x7f4a8f3836c0 pvol_id = 391 logic_id = 308 
state: navInd=109644, id=3536, level=2/3,  onBoundary=true, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000/SFBY0x7f4a8f3836c0>
Distance from boundary point: 1.064791535219556 [cm]
point in the middle before next crossing: (-23.578149423779063, -49.316102209046122, 2.613981083066565) [cm]
Volume at mid-distance: SFBY0x7f4a8f3836c0 pvol_id = 391 logic_id = 308 
state: navInd=109644, id=3536, level=2/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000/SFBY0x7f4a8f3836c0>
 consistency confirmed
```

amandalund commented:
Thanks so much @agheata for digging into this and for the detailed response!

@sethrj for the position and direction provided I'm actually not able to reproduce the failure either: reinitializing at `{-23.578149423779067,-49.316102209046122,2.6139810830665646}` gives volume SFBY, the same as @agheata is seeing. However, looking at some of the problematic cases in the SF geometry, I do see some that fail both the Celeritas VecGeom test:
```console
/home/alund/celeritas_project/celeritas/test/celeritas/GenericGeoTestBase.cc:330: Failure
Failed
reinitialization changed the volume at {-28.337719442113912,-44.948108144291496,2.9228566584671118} along {-0.78668992955890826,0.6136596936033234,0.067384977385729794} from SFLY1a to SF (alleged safety: 0.0092130674372086219)
```
and @agheata's test snippet:
```console
N9celeritas6detail12BVHNavigatorE_______________________________________________________________________________________
point: (-23.252307507414024, -48.914998036277758, 2.487258902170662) [cm] dir: (-0.786689929558908, 0.613659693603323, 0.067384977385730)  located in volume: SFBY0x7f4a8f3836c0 pvol_id = 391 logic_id = 308 
state: NavStateIndex: navInd=102576, level=2/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000/SFBY0x7f4a8f3836c0>
safety  = 0.000151003704627 [cm] distance = 6.452604357232426 [cm]
propagated point on boundary: (-28.328506374676707, -44.955294823475036, 2.922067500861830) [cm]
Relocated volume: SFLY1a0x7f4a8effa000 pvol_id = 886 logic_id = 313 
state: NavStateIndex: navInd=102112, level=1/3,  onBoundary=true, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000>
Distance from boundary point: 0.023422360172761 [cm]
point in the middle before next crossing: (-28.337719442113912, -44.948108144291496, 2.922856658467111) [cm]
Volume at mid-distance: SF0x7f4a8f07c780_PV pvol_id = 951 logic_id = 332 
state: NavStateIndex: navInd=1, level=0/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV>
 consistency confirmed
```
@agheata could you confirm that this is an inconsistency, starting at position `{-23.2523075074140237, -48.9149980362777583, 2.4872589021706619}` and direction `{-0.7866899295589083, 0.6136596936033234, 0.0673849773857298}`?

agheata commented:
> @agheata could you confirm that this is an inconsistency, starting at position `{-23.2523075074140237, -48.9149980362777583, 2.4872589021706619}` and direction `{-0.7866899295589083, 0.6136596936033234, 0.0673849773857298}`?

Yes @amandalund , I confirm that there is a consistency problem there and the point does not get correctly located inside `SFLY1a` volume where it actually is. I have traced the bug to a wrong bounding box extent calculated for `SFLY1`, which is a Boolean solid (I did not yet have the time to check its parameters). In these cases, the volume typically looks chopped when reached from outside, but entire when navigated from inside. I'll be looking for the bug, however, I have limited time since I'll be out of the office till the end of next week. Here is my debug output:

```
N3cxx12BVHNavigatorE_______________________________________________________________________________________
point: (-23.252307507414024, -48.914998036277758, 2.487258902170662) [cm] dir: (-0.786689929558908, 0.613659693603323, 0.067384977385730)  located in volume: SFBY0x7f4a8f3836c0 pvol_id = 391 logic_id = 308 
state: NavStateIndex: navInd=102576, level=2/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000/SFBY0x7f4a8f3836c0>
safety  = 0.000151003704627 [cm] distance = 6.452604357232424 [cm]
propagated point on boundary: (-28.328506374676703, -44.955294823475036, 2.922067500861830) [cm]
Relocated volume: SFLY1a0x7f4a8effa000 pvol_id = 886 logic_id = 313 
state: NavStateIndex: navInd=102112, level=1/3,  onBoundary=true, path=</SF0x7f4a8f07c780_PV/SFLY1a0x7f4a8effa000>
Distance from boundary point: 0.023422360172764 [cm]
point in the middle before next crossing: (-28.337719442113912, -44.948108144291496, 2.922856658467111) [cm]
middle point NOT contained in the bounding box of the relocated volume (Extent?)
middle point in local solid coordinates: (5.971891855708503, 11.534794498894311, 0.424259192357652) [cm]
Local extent: amin= (-11.544007566331523, -12.299604669236672, -6.505683058191558) [cm], amax= (16.757924943219571, 9.145954024215715, 6.693394173661474) [cm]
Volume at mid-distance: SF0x7f4a8f07c780_PV pvol_id = 951 logic_id = 332 
state: NavStateIndex: navInd=1, level=0/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV>
 INCONSISTENT location
```

sethrj commented:
Yay! I bet that's causing the rest of our consistency issues. Thanks for finding and reproducing the failure @amandalund and @agheata !

agheata commented:
Note that this critical bug is fixed now in the master by https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/991 which is also merged in the v1-patches.

sethrj committed:
Use latest fixes in vecgeom

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vecgeom-sf-failure

sethrj commented:
@agheata Thanks a bunch! With that fix, we now have completely consistent results with and without using the safety distance to cross boundaries in CMS 2018! (see #844 ). Perhaps you/Gabriele can tag that VecGeom 1.2.5?

agheata commented:
> @agheata Thanks a bunch! With that fix, we now have completely consistent results with and without using the safety distance to cross boundaries in CMS 2018! (see #844 ). Perhaps you/Gabriele can tag that VecGeom 1.2.5?

Done

sethrj committed:
Update vecgeom to v1.2.5

sethrj committed:
Update unit tests

sethrj committed:
Remove sf test

sethrj committed:
Fix tests

amandalund commented:
Thanks so much @agheata! I've confirmed that with this fix we no longer encounter the problem above.

Using VecGeom 1.2.5 the number of volume mismatches is again significantly reduced; however, I am still seeing some inconsistencies. Using the same SF geometry and tests as above, the next problem I run into is with starting position `{-20.4418060929454626, 59.9848538921079779, 1.1899989221172371}` and direction `{0.4128738348137703, 0.9043875740053724, 0.1077882763144891}`:
```console
/home/alund/celeritas_project/celeritas/test/celeritas/GenericGeoTestBase.cc:348: Failure
Failed
reinitialization changed the volume at {-20.438346754379019,59.992431467991857,1.1909020458073922} along {0.41287383481377032,0.9043875740053724,0.1077882763144891} from SFAbs2PbLinPb to SFAbs2PbLOutAl (alleged safety: 0.00090204580739225437)      
```
and
```console
N9celeritas6detail12BVHNavigatorE_______________________________________________________________________________________
point: (-20.441806092945463, 59.984853892107978, 1.189998922117237) [cm] dir: (0.412873834813770, 0.904387574005372, 0.107788276314489)  located in volume: SFAbs2Al10x7f4a8f03dd80 pvol_id = 688 logic_id = 281                                                      
state: NavStateIndex: navInd=2144, level=1/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFAbs2Al10x7f4a8f03dd80>
safety  = 0.000001077882763 [cm] distance = 0.000009999999998 [cm]                     
propagated point on boundary: (-20.441801964207116, 59.984862935983713, 1.190000000000000) [cm]
Relocated volume: SFAbs2PbLinPb0x7f4a8f03df00 pvol_id = 689 logic_id = 282             
state: NavStateIndex: navInd=2172, level=1/3,  onBoundary=true, path=</SF0x7f4a8f07c780_PV/SFAbs2PbLinPb0x7f4a8f03df00>
Distance from boundary point: 0.016737363992353 [cm]                                   
point in the middle before next crossing: (-20.438346754379019, 59.992431467991857, 1.190902045807392) [cm]
Volume at mid-distance: SFAbs2PbLOutAl0x7f4a8f03ec40 pvol_id = 691 logic_id = 297      
state: NavStateIndex: navInd=2228, level=1/3,  onBoundary=false, path=</SF0x7f4a8f07c780_PV/SFAbs2PbLOutAl0x7f4a8f03ec40>
```

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vecgeom-sf-failure

sethrj committed:
REVERTME: field propagator debug code

This reverts commit be1b1b26a2468e7bd3fe5593086368f9738f89b6.

sethrj committed:
Fix boundary inconsistency with checked geo track view and GeantGeo

sethrj committed:
Fix erroneous condition due to #875

The `remaining > driver_.minimum_step` condition was changed to
`remaining > 0`, but the "halving due to boundary crossing issues" was
not updated to account for the fact that the *input* remaining step
distance was less than the tolerance, so it would get stuck in a loop of
continually halving.

sethrj committed:
Revert "REVERTME: field propagator debug code"

This reverts commit 84d1e753b140f6fee0399d1a77e37397c64ccbeb.

sethrj committed:
Update geant4 field propagator results

sethrj commented:
I'm glad I added these tests; the combination of the checked geometry and new testing caught a subtle bug in the field propagation near boundaries.

sethrj committed:
fixup! Update geant4 field propagator results

agheata commented:
> `{-20.4418060929454626, 59.9848538921079779, 1.1899989221172371}` and direction `{0.4128738348137703, 0.9043875740053724, 0.1077882763144891}`

@amandalund I confirm the inconsistency for this point, but this is due to an overlap. Actually replacing the BVHNavigator with the LoopNavigator the inconsistency disappears, just because by chance the order of checking the two candidate volumes is different compared with what the BVH does. I have checked that the point is actually contained in both: `/SF0x7f4a8f07c780_PV/SFAbs2PbLinPb0x7f4a8f03df00` and `/SF0x7f4a8f07c780_PV/SFAbs2PbLOutAl0x7f4a8f03ec40`, and confirmed by importing the geometry in ROOT and running the sampling-based overlap checker. The checker gives the picture below, where `SFAbs2PbLinPb` and `SFAbs2PbLOutAl` are the green and blue volumes, and there are some red dots at the boundary between them, which are the points sampled to be contained in both. 

To reproduce at the ROOT prompt:
```console
root[0] TGeoManager::Import("sf.gdml");
root [1] gGeoManager->CheckOverlaps(0.01, "s10000000");
root[2] gGeoManager->PrintOverlaps();
=== Overlaps for Default ===
 = Overlap ov00000: Volume SF: node SFAbs1PbLinPb0x7f4a8f038540 overlapping SFAbs1PbLOutAl0x7f4a8f03d380 ovlp=0.00991886
 = Overlap ov00001: Volume SF: node SFAbs2PbLinPb0x7f4a8f03df00 overlapping SFAbs2PbLOutAl0x7f4a8f03ec40 ovlp=0.0096247
root[3] auto overlap = (TGeoOverlap*)gGeoManager->GetListOfOverlaps()->At(0);
root[4] overlap->Draw("ogl"); // then move the mouse in the GL window since the volumes are thin
```

![overlap_  SFAbs1PbLinPb_SFAbs1PbLOutAl](https://github.com/celeritas-project/celeritas/assets/18400453/6572735f-9711-43b1-afe6-f59f58d57a20)

agheata commented:
> I'm glad I added these tests; the combination of the checked geometry and new testing caught a subtle bug in the field propagation near boundaries.

@seth obviously we are lacking some generic GDML-based integration tests for consistency similar to this one in VecGeom, we have several solid-based ones but very few involving global navigation. So if one of you will find some time to add such a test directly in VecGeom, it would be great.

sethrj commented:
Thanks again @agheata . It's essentially impossible to remove all overlaps from all geometries, so we'll have to make sure our navigator is robust against small rare overlaps. I think it is (it doesn't seem to crash or do anything too unusual), but the extra checks @amandalund added highlight the inconsistency.

sethrj committed:
Improve test skipping logic in failing test

sethrj committed:
Make tests less verbose by default

This makes skipped tests easier to find.

sethrj committed:
Hide 'running in serial'

sethrj commented:
>  Obviously we are lacking some generic GDML-based integration tests for consistency similar to this one in VecGeom, we have several solid-based ones but very few involving global navigation. So if one of you will find some time to add such a test directly in VecGeom, it would be great.

Before we add such tests, perhaps we need to move the BVH navigator into VecGeom from AdePT (and remove our Celeritas copy of it)?

sethrj committed:
Make vecgeom failure testing more robust

sethrj committed:
fixup! Make vecgeom failure testing more robust

agheata commented:
> Before we add such tests, perhaps we need to move the BVH navigator into VecGeom from AdePT (and remove our Celeritas copy of it)?

Yes, we should do that first, now that the navigator is quite stable.

amandalund commented:
Thanks again for your help @agheata and for the tip on checking for overlaps -- I'm pretty sure now that all of the inconsistencies I'm seeing in the SF geometry (and most of the ones in the full cms2018) are due to overlaps.

sethrj commented:
@amandalund *finally* all the tests are passing... if you can find time to review this today (I know it's on your radar!) we can merge this and release v0.3.1.

amandalund commented on test/celeritas/ext/Vecgeom.test.cc:
```diff
@@ -774,6 +774,91 @@ TEST_F(SolidsTest, reflected_vol)
+            403.9, 650, 650, 403.9, 549.15, 1096.95, 11200, 9.9999999999992,
+            180, 910, 24000, 6000};
+        EXPECT_VEC_SOFT_EQ(expected_distances, result.distances);
+        static real_type const expected_hw_safety[] = {100, 2.1499999999997,
+            0.52499999999986, 13.023518051922, 6.95, 6.95, 13.023518051922,
+            0.52499999999986, 2.15, 100, 5, 8, 100, 100, 100};
```
I guess we're seeing a much more conservative safety for vecgeom because the `BEAM`s are polycones?

sethrj replied:
Correct. This is the bug where the safety is limited to the virtual planes of the polycone: https://github.com/celeritas-project/celeritas/pull/844#issuecomment-1636815806 and https://sft.its.cern.ch/jira/browse/VECGEOM-618

amandalund reviewed (approved):
This looks great @sethrj! Thanks for the extra testing -- I like the `CheckedGeoTrackView`. Should be (and already has been!) very helpful.
